pkg/controller/constrainttemplate/constrainttemplate_controller.go

```diff
@@ -848,3 +737,174 @@ func v1beta1ToV1(v1beta1Obj *admissionregistrationv1beta1.ValidatingAdmissionPol
+			return err
+		}
+
+		if currentVap == nil {
+			logger.Info("creating VAP", "vapName", vapName)
+			if err := r.Create(ctx, newVap); err != nil {
+				logger.Info("creating VAP error", "vapName", vapName, "error", err)
+				err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not create VAP object", status, err)
+				return err
+			}
+			// after VAP is created, trigger update event for all constraints
+			if err := r.triggerConstraintEvents(ctx, ct, status); err != nil {
+				return err
+			}
+		} else if !reflect.DeepEqual(currentVap, newVap) {
+			logger.Info("updating VAP")
+			if err := r.Update(ctx, newVap); err != nil {
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not update VAP object", status, err)
+				return err
+			}
+		}
+	}
+	// do not generate VAP resources
+	// remove if exists
+	if !generateVap && isVapAPIEnabled && groupVersion != nil {
+		currentVap, err := vapForVersion(groupVersion)
+		if err != nil {
+			logger.Error(err, "error getting VAP object with respective groupVersion")
+			err := r.reportErrorOnCTStatus(ctx, ErrCreateCode, "Could not get VAP with correct group version", status, err)
+			return err
+		}
+		vapName := fmt.Sprintf("gatekeeper-%s", unversionedCT.GetName())
+		logger.Info("check if VAP exists", "vapName", vapName)
+		if err := r.Get(ctx, types.NamespacedName{Name: vapName}, currentVap); err != nil {
+			if !apierrors.IsNotFound(err) && !errors.As(err, &discoveryErr) && !meta.IsNoMatchError(err) {
+				return err
+			}
+			currentVap = nil
+		}
+		if currentVap != nil {
+			logger.Info("deleting VAP")
+			if err := r.Delete(ctx, currentVap); err != nil {
+				err := r.reportErrorOnCTStatus(ctx, ErrUpdateCode, "Could not delete VAP object", status, err)
+				return err
+			}
+			// after VAP is deleted, trigger update event for all constraints
+			if err := r.triggerConstraintEvents(ctx, ct, status); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+// updateTemplateWithBlockVAPBGenerationAnnotations updates the ConstraintTemplate with an annotation to block VAPB generation until specific time
+// This is to avoid the issue where the VAPB is generated before the CRD is cached in the API server.
+func (r *ReconcileConstraintTemplate) updateTemplateWithBlockVAPBGenerationAnnotations(ctx context.Context, ct *v1beta1.ConstraintTemplate) error {
+	currentTime := time.Now()
+	if ct.Annotations != nil && ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation] != "" {
+		until := ct.Annotations[constraint.BlockVAPBGenerationUntilAnnotation]
+		t, err := time.Parse(time.RFC3339, until)
+		if err != nil {
+			return err
+		}
+		// if wait time is within the time window to generate vap binding, do not update the annotation
+		// otherwise update the annotation with the current time + wait time. This protects against manual updates on annotations with a timestamp that prevents binding from getting generated.
```

Comment:
I think, more accurately, this prevents clock skew from preventing generation on task reschedule.

This design really doesn't protect against malicious users (RBAC for templates does that)